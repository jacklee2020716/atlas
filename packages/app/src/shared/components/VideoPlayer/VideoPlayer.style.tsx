import { css } from '@emotion/core'
import { breakpoints } from '../../theme'

export type VideoPlayerStyleProps = {
  width?: string | number
  height?: string | number
  responsive?: boolean
  ratio?: string
}

export const makeStyles = ({ width = '100%', height = '100%' }: VideoPlayerStyleProps) => {
  return {
    containerStyles: css`
      max-width: ${breakpoints.medium};
      & .video-player {
      }
    `,
    playerStyles: css`
      width: ${width};
      height: ${height};
    `,
  }
}
