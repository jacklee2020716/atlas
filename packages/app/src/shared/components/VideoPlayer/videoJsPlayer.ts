import videojs, { VideoJsPlayer, VideoJsPlayerOptions } from 'video.js'
import { RefObject, useEffect, useRef, useState } from 'react'
import 'video.js/dist/video-js.css'

export type VideoJsConfig = {
  src: string
  width?: number
  height?: number
  fluid?: boolean
  fill?: boolean
}

type VideoJsPlayerHook = (config: VideoJsConfig) => [VideoJsPlayer | null, RefObject<HTMLVideoElement>]
export const useVideoJsPlayer: VideoJsPlayerHook = ({ fill, fluid, height, src, width }) => {
  const playerRef = useRef<HTMLVideoElement>(null)
  const [player, setPlayer] = useState<VideoJsPlayer | null>(null)

  useEffect(() => {
    const videoJsOptions: VideoJsPlayerOptions = {
      controls: true,
    }

    const playerInstance = videojs(playerRef.current, videoJsOptions)
    setPlayer(playerInstance)

    return () => {
      playerInstance.dispose()
    }
  }, [])

  useEffect(() => {
    if (!player) {
      return
    }

    player.src({
      src,
      type: 'video/mp4',
    })
  }, [player, src])

  useEffect(() => {
    if (!player || !width) {
      return
    }

    player.width(width)
  }, [player, width])

  useEffect(() => {
    if (!player || !height) {
      return
    }

    player.height(height)
  }, [player, height])

  useEffect(() => {
    if (!player) {
      return
    }

    player.fluid(Boolean(fluid))
  }, [player, fluid])

  useEffect(() => {
    if (!player) {
      return
    }

    // @ts-ignore @types/video.js is outdated and doesn't provide types for some newer video.js features
    player.fill(Boolean(fill))
  }, [player, fill])

  return [player, playerRef]
}
