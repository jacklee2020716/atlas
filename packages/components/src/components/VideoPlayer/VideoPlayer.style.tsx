import { css } from "@emotion/core";
import { breakpoints } from "../../theme";

export type VideoPlayerStyleProps = {
	width?: string | number;
	height?: string | number;
	responsive?: boolean;
	ratio?: string;
};

export let makeStyles = ({
	width = "100%",
	height = "100%",
	responsive = true,
	ratio = "16:9",
}: VideoPlayerStyleProps) => {
	let ratioPerc = ratio
		.split(":")
		.map(x => Number(x))
		.reduce((x, y) => (y / x) * 100);

	return {
		containerStyles: css`
			max-width: ${breakpoints.medium};
			& .video-player {
			}
		`,
		playerStyles: css`
			width: ${width};
			height: ${height};
		`,
	};
};
