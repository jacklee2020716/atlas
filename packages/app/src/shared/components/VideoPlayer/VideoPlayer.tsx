import React from 'react'
import ReactPlayer from 'react-player'
import { VideoPlayerStyleProps, makeStyles } from './VideoPlayer.style'

export type VideoPlayerProps = {
  src?: string
  playing?: boolean
  poster?: string
  controls?: boolean
  volume?: number
  loop?: boolean
  autoPlay?: boolean
  muted?: boolean
  className?: string
  onReady?(): void
  onStart?(): void
  onPlay?(): void
  onPause?(): void
  onBuffer?(): void
  onEnded?(): void
  onError?(error: any): void
  onDuration?(duration: number): void
  onProgress?(state: { played: number; loaded: number }): void
} & VideoPlayerStyleProps

export default function VideoPlayer({
  src,
  poster,
  playing,
  autoPlay,
  loop = false,
  onStart,
  onReady,
  onPlay,
  onBuffer,
  onEnded,
  onDuration,
  onProgress,
  className,
  controls = true,
  ...styleProps
}: VideoPlayerProps) {
  const { playerStyles, containerStyles } = makeStyles(styleProps)
  return (
    <div css={containerStyles}>
      <ReactPlayer
        css={playerStyles}
        width={styleProps.responsive ? '100%' : styleProps.width}
        height={styleProps.responsive ? '100%' : styleProps.height}
        url={src}
        autoPlay={autoPlay}
        light={poster || true}
        className={className}
        playing={playing}
        loop={loop}
        controls={controls}
        onStart={onStart}
        onPlay={onPlay}
        onBuffer={onBuffer}
        onReady={onReady}
        onEnded={onEnded}
        onDuration={onDuration}
        onProgress={onProgress}
        config={{
          file: {
            attributes: {
              className: 'video-player',
            },
          },
        }}
      />
    </div>
  )
}
