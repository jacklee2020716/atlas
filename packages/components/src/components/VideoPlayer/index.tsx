import { memo } from "react"
import VideoPlayer from "./VideoPlayer"
export default memo(VideoPlayer)
