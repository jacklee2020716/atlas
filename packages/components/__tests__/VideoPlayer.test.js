import React from "react"
import { mount } from "enzyme"
import VideoPlayer from "./../src/components/VideoPlayer"

describe("VideoPlayer component", () => {

  const component = mount(
    <VideoPlayer
      src={"https://www.computt.com/asset/v0/5EjgEKNpyDbNdjcJoZ8izWuzeDUtAcaUvwG8vUWZQZ256NLb"}
      poster={"https://ssl-static.libsyn.com/p/assets/a/4/8/f/a48f1a0697e958ce/Cover_2.png"}
    />)

  it("Should render correctly", () => {
    expect(component).toMatchSnapshot()
  })

})