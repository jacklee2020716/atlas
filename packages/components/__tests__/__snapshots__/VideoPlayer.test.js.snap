// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer component Should render correctly 1`] = `
<Memo(VideoPlayer)
  poster="https://ssl-static.libsyn.com/p/assets/a/4/8/f/a48f1a0697e958ce/Cover_2.png"
  src="https://www.computt.com/asset/v0/5EjgEKNpyDbNdjcJoZ8izWuzeDUtAcaUvwG8vUWZQZ256NLb"
>
  <div
    css={
      Object {
        "map": undefined,
        "name": "19dglr4",
        "next": undefined,
        "styles": "
      max-width: 48rem;
      & .video-player {
      }
    ",
        "toString": [Function],
      }
    }
  >
    <ReactPlayer
      config={
        Object {
          "file": Object {
            "attributes": Object {
              "className": "video-player",
            },
          },
        }
      }
      controls={true}
      css={
        Object {
          "map": undefined,
          "name": "bggj8i",
          "next": undefined,
          "styles": "
      width: 100%;
      height: 100%;
    ",
          "toString": [Function],
        }
      }
      height="360px"
      light="https://ssl-static.libsyn.com/p/assets/a/4/8/f/a48f1a0697e958ce/Cover_2.png"
      loop={false}
      muted={false}
      onBuffer={[Function]}
      onBufferEnd={[Function]}
      onDisablePIP={[Function]}
      onDuration={[Function]}
      onEnablePIP={[Function]}
      onEnded={[Function]}
      onError={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      onProgress={[Function]}
      onReady={[Function]}
      onSeek={[Function]}
      onStart={[Function]}
      pip={false}
      playbackRate={1}
      playing={false}
      playsinline={false}
      progressInterval={1000}
      style={Object {}}
      url="https://www.computt.com/asset/v0/5EjgEKNpyDbNdjcJoZ8izWuzeDUtAcaUvwG8vUWZQZ256NLb"
      volume={null}
      width="640px"
      wrapper="div"
    >
      <div
        css={
          Object {
            "map": undefined,
            "name": "bggj8i",
            "next": undefined,
            "styles": "
      width: 100%;
      height: 100%;
    ",
            "toString": [Function],
          }
        }
        style={
          Object {
            "height": "360px",
            "width": "640px",
          }
        }
      >
        <Preview
          light="https://ssl-static.libsyn.com/p/assets/a/4/8/f/a48f1a0697e958ce/Cover_2.png"
          onClick={[Function]}
          url="https://www.computt.com/asset/v0/5EjgEKNpyDbNdjcJoZ8izWuzeDUtAcaUvwG8vUWZQZ256NLb"
        >
          <div
            className="react-player__preview"
            onClick={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundImage": "url(https://ssl-static.libsyn.com/p/assets/a/4/8/f/a48f1a0697e958ce/Cover_2.png)",
                "backgroundPosition": "center",
                "backgroundSize": "cover",
                "cursor": "pointer",
                "display": "flex",
                "height": "100%",
                "justifyContent": "center",
                "width": "100%",
              }
            }
          >
            <div
              className="react-player__shadow"
              style={
                Object {
                  "alignItems": "center",
                  "background": "radial-gradient(rgb(0, 0, 0, 0.3), rgba(0, 0, 0, 0) 60%)",
                  "borderRadius": "64px",
                  "display": "flex",
                  "height": "64px",
                  "justifyContent": "center",
                  "width": "64px",
                }
              }
            >
              <div
                className="react-player__play-icon"
                style={
                  Object {
                    "borderColor": "transparent transparent transparent white",
                    "borderStyle": "solid",
                    "borderWidth": "16px 0 16px 26px",
                    "marginLeft": "7px",
                  }
                }
              />
            </div>
          </div>
        </Preview>
      </div>
    </ReactPlayer>
  </div>
</Memo(VideoPlayer)>
`;
